pkg/cloud/cloud.go

```diff
@@ -855,26 +861,60 @@ func (c *cloud) ResizeDisk(ctx context.Context, volumeID string, newSizeBytes in
 	newSizeGiB := util.RoundUpGiB(newSizeBytes)
 	oldSizeGiB := aws.Int64Value(volume.Size)
 
+	latestMod, modFetchError := c.getLatestVolumeModification(ctx, volumeID)
+
+	if latestMod != nil && modFetchError == nil {
+		state := aws.StringValue(latestMod.ModificationState)
+		if state == ec2.VolumeModificationStateModifying {
+			return oldSizeGiB, fmt.Errorf("volume %q is still being expanded to size %d", volumeID, newSizeGiB)
```

Comment:
In order to make this function idempotent, we shouldn't return an error if the volume modification is in progress, no?

It feels like it should return success _if the volume is already being modified to newSizeGiB_, but the volume is being modified to a value _different than newSizeGiB_, then it's OK to return an error?

Reply:
that doesn't seem right either, then it would return that resize has succeeded which is a lie. 

attach/controllerpublish behaviour is to wait for the attach to complete if it's found to be in progress. The same could be done here

Reply:
Yep - was just about to push a change that made the whole RPC call block until operation is finished.

Reply:
> that doesn't seem right either, then it would return that resize has succeeded which is a lie.

That's a good point, but currently we're doing the same thing in the first call (because it considers the volume resized once the operation is in the Optimizing state). So if Optimizing mean success for the first call, it should also mean success for the subsequent calls.

> attach/controllerpublish behaviour is to wait for the attach to complete if it's found to be in progress. The same could be done here

+1

> Yep - was just about to push a change that made the whole RPC call block until operation is finished.

By waiting for the modification to be in the Completed state? If so, can/should we change the behaviour in-tree as well, to be consistent with the CSI driver?

Reply:
Alright modified so that each subsequent resize call will now wait for previous resize to finish before returning. 

> By waiting for the modification to be in the Completed state? If so, can/should we change the behaviour in-tree as well, to be consistent with the CSI driver?

No not "Completed" state. I reverted the code to "optimizing" *and* "completed" check, because the main reason this was breaking was because `describeVolume` can return updated size before volume has really finished resizing. "optimized" vs "completed" state had no impact on the outcome.

Reply:
double checking, doc says size can take a few seconds after volume is in optimizing state, and I guess it does not matter because even if fs resize is attempted prematurely, the fs resize will just get retried and must eventually succeed.

"Size changes usually take a few seconds to complete and take effect after a volume is in the Optimizing state. "
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-volume-modifications.html

Reply:
We don't return size in fs resize, it just blindly tries to resize to the full disk size, whatever it may be at the time. So is it possible that we won't fs expand to the intended size if we don't wait?

Reply:
shoot you are right. https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/9b33867eb74b03f6525166cc84c4b68132aa9278/vendor/k8s.io/kubernetes/pkg/util/resizefs/resizefs_linux.go#L38

I am probably misreading the doc, the correct meaning must be that "optimizing" means the size change has already taken effect and only the performance change is still in progress.

Reply:
> Size changes usually take a few seconds to complete and take effect after a volume is in the Optimizing state.

tbh - the wording here is imprecise in EBS docs. We have been told by EBS engineers that volume is "ready-to-use"after it enters "optimizing" state but if we wait for "VolumeModificationComplete" state then, it can take really long time to complete resize operation and afaict there is no intermediate state between "optimizing" and "complete" (although docs do imply that there is a intermediate state where size changes take effect few seconds after volume enters "optimizing" state).